src/jit/codegenarm64.cpp

```diff
@@ -2665,62 +2665,191 @@ void CodeGen::genJumpTable(GenTree* treeNode)
-    }
     switch (treeNode->OperGet())
     {
-    case GT_LOCKADD:
-        instGen(INS_lock);
-        ins = INS_add;
-        break;
-    case GT_XCHG:
-        // lock is implied by xchg
-        ins = INS_xchg;
-        break;
-    case GT_XADD:
-        instGen(INS_lock);
-        ins = INS_xadd;
-        break;
-    default:
-        unreached();
+        case GT_XADD:
+        case GT_LOCKADD:
+            if (data->isContainedIntOrIImmed())
+            {
+                // Even though INS_add is specified here, the encoder will choose either
+                // an INS_add or an INS_sub and encode the immediate as a positive value
+                genInstrWithConstant(INS_add, emitActualTypeSize(treeNode), storeDataReg, loadReg,
+                                     data->AsIntConCommon()->IconValue(), REG_NA);
+            }
+            else
+            {
+                getEmitter()->emitIns_R_R_R(INS_add, emitActualTypeSize(treeNode), storeDataReg, loadReg, dataReg);
+            }
+            break;
+        case GT_XCHG:
+            assert(!data->isContained());
+            storeDataReg = dataReg;
+            break;
+        default:
+            unreached();
     }
-    getEmitter()->emitInsBinary(ins, emitActualTypeSize(data), &i, data);
+
+    getEmitter()->emitIns_R_R_R(INS_stlxr, emitTypeSize(treeNode), exResultReg, storeDataReg, addrReg);
+
+    getEmitter()->emitIns_J_R(INS_cbnz, EA_4BYTE, labelRetry, exResultReg);
+
+    gcInfo.gcMarkRegSetNpt(addr->gtGetRegMask());
 
     if (treeNode->gtRegNum != REG_NA)
     {
         genProduceReg(treeNode);
     }
-#else  // !0
-    NYI("genLockedInstructions");
-#endif // !0
+}
+
+//------------------------------------------------------------------------
+// genCodeForSwap: Produce code for a GT_CMPXCHG node.
+//
+// Arguments:
+//    tree - the GT_CMPXCHG node
+//
+void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* treeNode)
+{
+    assert(treeNode->OperIs(GT_CMPXCHG));
+
+    GenTreePtr addr      = treeNode->gtOpLocation;  // arg1
+    GenTreePtr data      = treeNode->gtOpValue;     // arg2
+    GenTreePtr comparand = treeNode->gtOpComparand; // arg3
+
+    regNumber targetReg    = treeNode->gtRegNum;
+    regNumber dataReg      = data->gtRegNum;
+    regNumber addrReg      = addr->gtRegNum;
+    regNumber comparandReg = comparand->gtRegNum;
+    regNumber exResultReg  = treeNode->ExtractTempReg(RBM_ALLINT);
+
+    // Check allocator assumptions
+    assert(addrReg != targetReg);
+    assert(dataReg != targetReg);
+    assert(comparandReg != targetReg);
+    assert(addrReg != dataReg);
+    assert(targetReg != REG_NA);
+    assert(exResultReg != REG_NA);
+    assert(exResultReg != targetReg);
+
+    assert(addr->isUsedFromReg());
+    assert(data->isUsedFromReg());
+    assert(!comparand->isUsedFromMemory());
+
+    // Store exclusive unpredictable cases must be avoided
+    assert(exResultReg != dataReg);
+    assert(exResultReg != addrReg);
```

Comment:
You might want some/all of these cases to be noway_asserts.
As assert()  only checks things in dbg/chk builds

Reply:
I changed them to asserts during development because I caught them failing and setting minopts

I set most of them back to noway_assert